File: src/main/java/core/procedures/math/trigonometry/Sinh.java
package core.procedures.math.trigonometry;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigComplex;
import core.scm.BigRatio;
import core.utils.Utils;

import java.math.BigDecimal;
import java.math.BigInteger;

public final class Sinh extends AFn {

  public Sinh() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Number.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "sinh";
  }

  @Override
  public Number apply1(Object arg) {
    /* Special cases */
    if (Utils.INSTANCE.isZero(arg)) {
      return 0L;
    }
    if (arg instanceof BigDecimal) {
      return sinh((BigDecimal)arg);
    } else if (arg instanceof BigInteger) {
      return sinh((BigInteger)arg);
    } else if (arg instanceof BigComplex) {
      return sinh((BigComplex)arg);
    } else if (arg instanceof BigRatio) {
      return sinh(((BigRatio)arg).toBigDecimal());
    }
    return Math.sinh(((Number)arg).doubleValue());
  }

  static double sinh(BigDecimal bd) {
    double v = bd.doubleValue();
    if (Double.isInfinite(v) || Double.isNaN(v)) {
      return Double.NaN;
    } else {
      return Math.sinh(v);
    }
  }

  static double sinh(BigInteger bi) {
    double v = bi.doubleValue();
    if (Double.isInfinite(v) || Double.isNaN(v)) {
      return Double.NaN;
    } else {
      return Math.sinh(v);
    }
  }

  /* sinh(x + yi) = sinh(x)*cos(y) + cosh(x)*sin(y)*i */
  static Number sinh(BigComplex c) {
    BigDecimal x = c.getRe();
    BigDecimal y = c.getIm();
    double re = sinh(x) * Cos.cos(y);
    double im = Cosh.cosh(x) * Sin.sin(y);
    if (Double.isInfinite(re) || Double.isNaN(re)) {
      return Double.NaN;
    }
    if (Double.isInfinite(im) || Double.isNaN(im)) {
      return Double.NaN;
    }
    return new BigComplex(re, im);
  }
}

File: src/main/java/core/procedures/math/trigonometry/Sinh.kt
package core.procedures.math.trigonometry

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Complex
import core.utils.Utils

import kotlin.math.cos
import kotlin.math.cosh
import kotlin.math.sin
import kotlin.math.sinh

class Sinh : AFn<Number?, Number>(name = "sinh", isPure = true, arity = Exactly(1),
                                  mandatoryArgsTypes = arrayOf(Number::class.java)) {

    override operator fun invoke(arg: Number?) = when {
        Utils.isZero(arg) -> 0L
        arg is Complex -> sinh(arg)
        else              -> sinh(arg!!.toDouble())
    }

    companion object {
        /* sinh(x + yi) = sinh(x)*cos(y) + cosh(x)*sin(y)*i */
        internal fun sinh(c: Complex): Number {
            val x = c.re.toDouble()
            val y = c.im.toDouble()
            val re = sinh(x) * cos(y)
            if (!re.isFinite()) {
                return re
            }
            val im = cosh(x) * sin(y)
            if (!im.isFinite()) {
                return im
            }
            return Complex(re, im)
        }
    }
}
